File: trunk/examples/zoo/mesh/mesh.m

% Open an XML run file to write into
lib = rpLib(infile);

meshtype = rpLibGetString(lib, 'input.choice(mesh).current');
contour = rpLibGetString(lib, 'input.boolean(contour).current')
switch contour 
  case 'yes'
    view = 'contour';
  otherwise
    view = 'heightmap';
end

switch meshtype 
  case 'cloud' 
    mesh = 'output.mesh';
    rpLibPutString(lib, 'output.mesh.about.label', \
		   'cloud in unstructured mesh', 0);
    rpLibPutString(lib, 'output.mesh.dim', '2', 0);
    rpLibPutString(lib, 'output.mesh.units', 'm', 0);
    rpLibPutString(lib, 'output.mesh.hide', 'yes', 0);

    xv = linspace(0, 1, 50)
    for i = 1:50
      for j = 1:50
        string = sprintf('%12g  %12g\n', xv(i), xv(j));
	rpLibPutString(lib, 'output.mesh.unstructured.points', string, 1);
      endfor
    endfor
    rpLibPutString(lib, 'output.mesh.unstructured.celltypes', '', 0);

  case 'regular' 
    mesh = 'output.mesh';
    rpLibPutString(lib, 'output.mesh.about.label', 'uniform grid mesh', 0);
    rpLibPutString(lib, 'output.mesh.dim', '2', 0);
    rpLibPutString(lib, 'output.mesh.units', 'm', 0);
    rpLibPutString(lib, 'output.mesh.hide', 'yes', 0);

    % Specify the x-axis of the mesh
    rpLibPutString(lib, 'output.mesh.grid.xaxis.min', '0.0', 0);
    rpLibPutString(lib, 'output.mesh.grid.xaxis.max', '1.0', 0);
    rpLibPutString(lib, 'output.mesh.grid.xaxis.numpoints', '50', 0);
    
    % Specify the y-axis of the mesh
    rpLibPutString(lib, 'output.mesh.grid.yaxis.min', '0.0', 0);
    rpLibPutString(lib, 'output.mesh.grid.yaxis.max', '1.0', 0);
    rpLibPutString(lib, 'output.mesh.grid.yaxis.numpoints', '50', 0);

  case 'irregular' 
    mesh = 'output.mesh';
    rpLibPutString(lib, 'output.mesh.about.label', 'irregular grid mesh', 0);
    rpLibPutString(lib, 'output.mesh.dim', '2', 0);
    rpLibPutString(lib, 'output.mesh.units', 'm', 0);
    rpLibPutString(lib, 'output.mesh.hide', 'yes', 0);

    xv = linspace(0, 1, 50)
    string = sprintf('%12g\n', xv);
    rpLibPutString(lib, 'output.mesh.grid.xcoords', string, 0);
    rpLibPutString(lib, 'output.mesh.grid.ycoords', string, 0);

  case 'hybrid' 
    mesh = 'output.mesh';
    rpLibPutString(lib, 'output.mesh.about.label', \
		   'hybrid regular and irregular grid mesh', 0);
    rpLibPutString(lib, 'output.mesh.dim', '2', 0);
    rpLibPutString(lib, 'output.mesh.units', 'm', 0);
    rpLibPutString(lib, 'output.mesh.hide', 'yes', 0);

    xv = linspace(0, 1, 50)
    string = sprintf('%g\n', xv);
    rpLibPutString(lib, 'output.mesh.grid.xcoords', string, 0);

    % Specify the y-axis of the mesh
    rpLibPutString(lib, 'output.mesh.grid.yaxis.min', '0.0', 0);
    rpLibPutString(lib, 'output.mesh.grid.yaxis.max', '1.0', 0);
    rpLibPutString(lib, 'output.mesh.grid.yaxis.numpoints', '50', 0);

  case 'triangular'
    mesh = 'output.mesh';
    rpLibPutString(lib, 'output.mesh.about.label', \
		   'triangles in unstructured mesh', 0);
    rpLibPutString(lib, 'output.mesh.dim', '2', 0);
    rpLibPutString(lib, 'output.mesh.units', 'm', 0);
    rpLibPutString(lib, 'output.mesh.hide', 'yes', 0);

    f = fopen('points.txt');
    points = fscanf(f, '%g');
    fclose(f);
    string = sprintf('%g %g\n', points);
    rpLibPutString(lib, 'output.mesh.unstructured.points', string, 0);

    rpLibPutFile(lib, 'output.mesh.unstructured.triangles', 'triangles.txt', \
		 0, 0);

  case 'unstructured' 
    mesh = 'output.mesh';
    rpLibPutString(lib, 'output.mesh.about.label', 'Unstructured Grid', 0);
    rpLibPutString(lib, 'output.mesh.dim', '2', 0);
    rpLibPutString(lib, 'output.mesh.units', 'm', 0);
    rpLibPutString(lib, 'output.mesh.hide', 'yes', 0);
    
    f = fopen('points.txt');
    points = fscanf(f, '%g');
    fclose(f);
    string = sprintf('%g %g\n', points);
    rpLibPutString(lib, 'output.mesh.unstructured.points', string, 0);

    rpLibPutFile(lib, 'output.mesh.unstructured.cells', 'triangles.txt', 0, 0);
    rpLibPutString(lib, 'output.mesh.unstructured.celltypes', 'triangle', 0);

  case 'cells' 
    mesh = 'output.mesh';
    rpLibPutString(lib, 'output.mesh.about.label', \
		   'unstructured grid with heterogeneous cells', 0);
    rpLibPutString(lib, 'output.mesh.dim', '2', 0);
    rpLibPutString(lib, 'output.mesh.units', 'm', 0);
    rpLibPutString(lib, 'output.mesh.hide', 'yes', 0);

    f = fopen('points.txt');
    points = fscanf(f, '%g');
    fclose(f);
    string = sprintf('%g %g\n', points);
    rpLibPutString(lib, 'output.mesh.unstructured.points', string, 0);

    rpLibPutFile(lib, 'output.mesh.unstructured.cells', 'triangles.txt', 0, 0);

    % TO BE COMPLETED
    rpLibPutString(lib, 'output.mesh.unstructured.celltypes', 'triangle', 0);

  case 'vtkmesh'
    mesh = 'output.mesh';
    rpLibPutString(lib, 'output.mesh.about.label', 'vtk mesh', 0);
    rpLibPutString(lib, 'output.mesh.dim', '2', 0);
    rpLibPutString(lib, 'output.mesh.units', 'm', 0);
    rpLibPutString(lib, 'output.mesh.hide', 'yes', 0);

    rpLibPutFile(lib, 'output.mesh.vtk', 'mesh.vtk', 0, 0);

  case 'vtkfield' 

    rpLibPutString(lib, 'output.field(substrate).about.label', \
	       'Substrate Surface', 0);
    rpLibPutFile(lib, 'output.field(substrate).component.vtk', 'file.vtk', \
		 0, 0);
    rpLibPutString(lib, 'output.string.current', '', 0);
    rpLibResult(lib);
    quit;

  otherwise
    error 'unknown mesh type $meshtype'
end

% Read in the data since we're not simulating anything...
f = fopen('substrate_data.txt');
substrate_values = fscanf(f, '%g');
fclose(f);
f = fopen('particle_data.txt');
particle_values = fscanf(f, '%g');
fclose(f);

rpLibPutString(lib, 'output.field(substrate).about.label', \
	       'Substrate Surface', 0);
rpLibPutString(lib, 'output.field(substrate).component.mesh', mesh, 0);
string = sprintf('%12g\n', substrate_values);
rpLibPutString(lib, 'output.field(substrate).component.values', string, 0);
rpLibPutString(lib, 'output.field(substrate).about.view', view, 0);

rpLibPutString(lib, 'output.field(particle).about.label', \
	       'Particle Surface', 0);
rpLibPutString(lib, 'output.field(particle).component.mesh', mesh, 0);
string = sprintf('%12g\n', particle_values);
rpLibPutString(lib, 'output.field(particle).component.values', string, 0);
rpLibPutString(lib, 'output.field(particle).about.view', view, 0);

rpLibPutString(lib, 'output.string.about.label', 'Mesh XML definition', 0);
xml = rpLibXml(lib);
rpLibPutString(lib, 'output.string.current', xml, 0);

% save the updated XML describing the run...
rpLibResult(lib);
quit;

File: examples/zoo/mesh/mesh.py
# ----------------------------------------------------------------------
#  EXAMPLE: Rappture <mesh> elements
# ======================================================================
#  AUTHOR:  Martin Hunt, Purdue University
#  Copyright (c) 2015  HUBzero Foundation, LLC
#
#  See the file "license.terms" for information on usage and
#  redistribution of this file, and for a DISCLAIMER OF ALL WARRANTIES.
# ======================================================================

import Rappture
import sys
import numpy as np

# uncomment these for debugging
# sys.stderr = open('mesh.err', 'w')
# sys.stdout = open('mesh.out', 'w')

rx = Rappture.PyXml(sys.argv[1])

meshtype = rx["input.choice(mesh).current"].value
contour = rx["input.boolean(contour).current"].value

if contour == "yes":
    view = 'contour'
else:
    view = 'heightmap'

mesh = rx['output.mesh']
mesh['dim'] = 2
mesh['units'] = "m"
mesh['hide'] = "yes"

x = np.linspace(0, 1, num=50, endpoint=True)
y = np.linspace(0, 1, num=50, endpoint=True)


if meshtype == 'cloud':
    mesh['about.label'] = "cloud in unstructured mesh"

    # Warning! Unstructured grids require newlines after each
    # set of points!.

    # use mgrid. weird syntax, but compact
    # x, y = np.mgrid[0:1:50j, 0:1:50j]
    # pts = (x.ravel('F'), y.ravel('F'))

    # OR create x and y vectors then use one of these methods

    # Use list comprehension to create string
    pts = ' '.join(["%s %s\n" % (a, b) for b in y for a in x])

    # Use meshgrid. Similar to mgrid()
    # X, Y = np.meshgrid(x, y)
    # pts = (X.ravel(), Y.ravel())

    mesh['unstructured.points'] = pts
    mesh['unstructured.celltypes'] = ""

if meshtype == 'regular':
    mesh['about.label'] = "uniform grid mesh"
    mesh['grid.xaxis.min'] = 0
    mesh['grid.xaxis.max'] = 1
    mesh['grid.xaxis.numpoints'] = 50
    mesh['grid.yaxis.min'] = 0
    mesh['grid.yaxis.max'] = 1
    mesh['grid.yaxis.numpoints'] = 50

if meshtype == 'irregular':
    mesh['about.label'] = "irregular grid mesh"
    mesh['grid.xcoords'] = x
    mesh['grid.ycoords'] = y

if meshtype == 'hybrid':
    mesh['about.label'] = "hybrid regular and irregular grid mesh"
    mesh['grid.xcoords'] = x
    mesh['grid.yaxis.min'] = 0
    mesh['grid.yaxis.max'] = 1
    mesh['grid.yaxis.numpoints'] = 50

if meshtype == 'triangular':
    mesh['about.label'] = "triangles in unstructured mesh"
    mesh.put('unstructured.points', 'points.txt', type='file')
    mesh.put('unstructured.triangles', 'triangles.txt', type='file')

if meshtype == 'unstructured':
    mesh['about.label'] = "Unstructured Grid"
    mesh['unstructured.celltypes'] = 'triangle'
    mesh.put('unstructured.points', 'points.txt', type='file')
    mesh.put('unstructured.cells', 'triangles.txt', type='file')

if meshtype == 'cells':
    # FIXME. Copied from tcl example, but looks just like 'unstructured'
    mesh['about.label'] = "Unstructured Grid with Heterogeneous Cells"
    mesh['unstructured.celltypes'] = 'triangle'
    mesh.put('unstructured.points', 'points.txt', type='file')
    mesh.put('unstructured.cells', 'triangles.txt', type='file')

if meshtype == 'vtkmesh':
    mesh['about.label'] = 'vtk mesh'
    mesh.put('vtk', 'mesh.vtk', type='file')

if meshtype == 'vtkfield':
    f = rx['output.field(substrate)']
    f['about.label'] = "Substrate Surface"
    f.put('component.vtk', 'file.vtk', type='file')
    rx.close()

f = rx['output.field(substrate)']
f['about.label'] = "Substrate Surface"
f['about.view'] = view
f['component.mesh'] = mesh.name
f.put('component.values', 'substrate_data.txt', type='file')

f = rx['output.field(particle)']
f['about.label'] = "Particle Surface"
f['about.view'] = view
f['component.mesh'] = mesh.name
f.put('component.values', 'particle_data.txt', type='file')

rx['output.string.about.label'] = "Mesh XML definition"
rx['output.string.current'] = mesh.xml()

rx.close()
